### PR TITLE
PivotGrid - Scroll position is wrong after expanding/collapsing node, if virtual mode is enabled (T984139)

### PR DESCRIPTION
**Pull request by novsstation**
Bugfix T984139

### CODE REVIEW

**Review by EugeniyKiyashko — DISMISSED** [review]

**Review by KuznetsovVN — DISMISSED** [review]

**Review by EugeniyKiyashko — DISMISSED** [review]

File: testing/tests/DevExpress.ui.widgets.pivotGrid/pivotGrid.tests.js [hunk 1]
```diff
@@ -2994,6 +2994,126 @@ QUnit.module('dxPivotGrid', {
...
+            }
+
+            function checkLeftTopVisibleHeaderCellTexts(pivotGrid, expected, message) {
+                const $rowsHeaderArea = pivotGrid.$element().find('.dx-pivotgrid-vertical-headers');
+                const $columnsHeaderArea = pivotGrid.$element().find('.dx-pivotgrid-horizontal-headers');
+
+                const rowsAreaRect = $rowsHeaderArea.get(0).getBoundingClientRect();
+                const columnsAreaRect = $columnsHeaderArea.get(0).getBoundingClientRect();
+
+                const expectedRowCellRect = getHeaderCellByText($rowsHeaderArea, expected.row.toString()).getBoundingClientRect();
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.pivotGrid/pivotGrid.tests.js]: expected.row.toString() -> expectedRowHeaderCellText ?

File: testing/tests/DevExpress.ui.widgets.pivotGrid/pivotGrid.tests.js [hunk 2]
```diff
@@ -2994,6 +2994,126 @@ QUnit.module('dxPivotGrid', {
...
 
+    ['row', 'column'].forEach((area) => {
+        [false, true].forEach((useNative) => {
+            function getHeaderCellByText($area, text) {
+                return $area.find('table:not(.dx-pivot-grid-fake-table) td')
+                    .filter((_, el) => { return el.innerText.trim() === text; })
+                    .get(0);
+            }
+
+            function checkLeftTopVisibleHeaderCellTexts(pivotGrid, expected, message) {
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.pivotGrid/pivotGrid.tests.js]: checkLeftTopVisibleHeaderCellTexts(pivotGrid -> imho it's clear what it means,
expected -> expectedRowHeaderCellText + expectedColHeaderCellText ?, 
message -> errorMessageDetails?
)

File: testing/tests/DevExpress.ui.widgets.pivotGrid/pivotGrid.tests.js [hunk 3]
```diff
@@ -2994,6 +2994,126 @@ QUnit.module('dxPivotGrid', {
...
+                const $rowsHeaderArea = pivotGrid.$element().find('.dx-pivotgrid-vertical-headers');
+                const $columnsHeaderArea = pivotGrid.$element().find('.dx-pivotgrid-horizontal-headers');
+
+                const rowsAreaRect = $rowsHeaderArea.get(0).getBoundingClientRect();
+                const columnsAreaRect = $columnsHeaderArea.get(0).getBoundingClientRect();
+
+                const expectedRowCellRect = getHeaderCellByText($rowsHeaderArea, expected.row.toString()).getBoundingClientRect();
+                const expectedColumnCellRect = getHeaderCellByText($columnsHeaderArea, expected.column.toString()).getBoundingClientRect();
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.pivotGrid/pivotGrid.tests.js]: expected.column.toString() -> expectedColHeaderCellText?

And, can we use 'elementFromPoint' instead?
Store {top, left} of the scrollable area and use elementFromPoint to get {top + 1, left + 1} to get the top/left visible cell:
document.elementFromPoint($('.dx-area-data-cell')[0].getBoundingClientRect().left + 1, $('.dx-area-data-cell')[0].getBoundingClientRect().top + 1).innerText - it's "$450" in ticket steps when "$2,410" is expected

File: testing/tests/DevExpress.ui.widgets.pivotGrid/pivotGrid.tests.js [hunk 4]
```diff
@@ -2994,6 +2994,126 @@ QUnit.module('dxPivotGrid', {
...
+
+                const expectedRow = 60;
+                const expectedColumn = area === 'row' ? 57 : 46;
+                checkLeftTopVisibleHeaderCellTexts(pivotGrid, { row: expectedRow, column: expectedColumn }, 'after scrolling');
+
+                const nodeToExpand = 65;
+                const getExpandedCells = () => pivotGrid.$element().find('.dx-pivotgrid-expanded');
+
+                pivotGrid.getDataSource().expandHeaderItem(area, [nodeToExpand]);
+                tryTriggerScrollEventAndWait(scrollable, this.clock);
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.pivotGrid/pivotGrid.tests.js]: Why test code should manually trigger 'scroll' after 'expandHeaderItem'?
Can test code call a method that executes all the necessary code?

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

File: testing/tests/DevExpress.ui.widgets.pivotGrid/pivotGrid.tests.js [hunk 5]
```diff
@@ -2994,6 +2994,126 @@ QUnit.module('dxPivotGrid', {
...
+
+                this.clock.tick(100);
+                checkLeftTopVisibleHeaderCellTexts(pivotGrid, { row: '1', column: '1' }, 'after initialization');
+
+                const scrollDistance = browser.msie ? 1950 : 1985; // there is a difference in font size for IE
+                const scrollable = pivotGrid._dataArea.groupElement().dxScrollable('instance');
+                scrollable.scrollTo({ left: scrollDistance, top: 2000 });
+                tryTriggerScrollEventAndWait(scrollable, this.clock);
+
+                const expectedRow = 60;
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.pivotGrid/pivotGrid.tests.js]: expectedRowHeaderCellText ?

File: testing/tests/DevExpress.ui.widgets.pivotGrid/pivotGrid.tests.js [hunk 6]
```diff
@@ -2994,6 +2994,126 @@ QUnit.module('dxPivotGrid', {
...
+                const expectedColumn = area === 'row' ? 57 : 46;
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.pivotGrid/pivotGrid.tests.js]: expectedColHeaderCellText?

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

**Review by EugeniyKiyashko — APPROVED** [review]

**Review by KuznetsovVN — APPROVED** [review]

**Review by IgnatovDan — APPROVED** [review]

**Pull request merged**